python/lsst/pipe/base/testUtils.py

```diff
@@ -61,29 +63,89 @@ def makeQuantum(task, butler, dataId, ioDataIds):
             ids = _normalizeDataIds(ioDataIds[name])
             for id in ids:
                 ref = _refFromConnection(butler, connection, id)
                 inputs[ref.datasetType].append(ref)
-        for name in connections.outputs:
+        except (ValueError, KeyError) as e:
+            raise ValueError(f"Error in connection {name}.") from e
+    for name in connections.outputs:
+        try:
             connection = connections.__getattribute__(name)
             _checkDataIdMultiplicity(name, ioDataIds[name], connection.multiple)
             ids = _normalizeDataIds(ioDataIds[name])
             for id in ids:
                 ref = _refFromConnection(butler, connection, id)
                 outputs[ref.datasetType].append(ref)
-        quantum = Quantum(taskClass=type(task),
-                          dataId=dataId,
-                          inputs=inputs,
-                          outputs=outputs)
-        return quantum
-    except KeyError as e:
-        raise ValueError("Mismatch in input data.") from e
+        except (ValueError, KeyError) as e:
+            raise ValueError(f"Error in connection {name}.") from e
+    quantum = Quantum(taskClass=type(task),
+                      dataId=dataId,
+                      inputs=inputs,
+                      outputs=outputs)
+    return quantum
+
+
+def _checkDimensionsMatch(universe, expected, actual):
+    """Test whether two sets of dimensions agree after conversions.
+
+    Parameters
+    ----------
+    universe : `lsst.daf.butler.DimensionUniverse`
+        The set of all known dimensions.
+    expected : `Set` [`str`] or `Set` [`~lsst.daf.butler.Dimension`]
+        The dimensions expected from a task specification.
+    actual : `Set` [`str`] or `Set` [`~lsst.daf.butler.Dimension`]
+        The dimensions provided by input.
+
+    Raises
+    ------
+    ValueError
+        Raised if ``expected`` and ``actual`` cannot be reconciled.
+    """
+    if _simplify(universe, expected) != _simplify(universe, actual):
+        raise ValueError(f"Mismatch in dimensions; expected {expected} but got {actual}.")
+
+
+def _simplify(universe, dimensions):
+    """Reduce a set of dimensions to a string-only form.
+
+    Parameters
+    ----------
+    universe : `lsst.daf.butler.DimensionUniverse`
+        The set of all known dimensions.
+    dimensions : `Set` [`str`] or `Set` [`~lsst.daf.butler.Dimension`]
+        A set of dimensions to simplify.
+
+    Returns
+    -------
+    dimensions : `Set` [`str`]
+        A copy of ``dimensions`` reduced to string form, with all spatial
+        dimensions simplified to ``skypix``.
+    """
+    simplified = set()
+    for dimension in dimensions:
+        # skypix not a real Dimension, handle it first
+        if dimension == "skypix":
+            simplified.add(dimension)
+        else:
+            # Need a Dimension to test spatialness
+            fullDimension = universe[dimension] if isinstance(dimension, str) else dimension
```

Comment:
I think at the very least that would fail formal type-checking (`DimensionUniverse.get` expects a string), and it might cause bugs depending on how `get` is implemented.